In [42]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf
import seaborn as sns
import scipy.stats as stats

**2.4**  
Table B.3 presents data on the gasoline mileage performance of 32 different automobiles.  
a. Fit a simple linear regression model relating gasoline mileage y (miles per gallon) to engine displacement xl (cubic inches).  

In [43]:
df = pd.read_csv('table_b3.csv')
df

y     x1   x2     x3    x4    x5  x6  x7     x8    x9   x10  x11
0   18.90  350.0  165  260.0  8.00  2.56   4   3  200.3  69.9  3910    1
1   17.00  350.0  170  275.0  8.50  2.56   4   3  199.6  72.9  3860    1
2   20.00  250.0  105  185.0  8.25  2.73   1   3  196.7  72.2  3510    1
3   18.25  351.0  143  255.0  8.00  3.00   2   3  199.9  74.0  3890    1
4   20.07  225.0   95  170.0  8.40  2.76   1   3  194.1  71.8  3365    0
5   11.20  440.0  215  330.0  8.20  2.88   4   3  184.5  69.0  4215    1
6   22.12  231.0  110  175.0  8.00  2.56   2   3  179.3  65.4  3020    1
7   21.47  262.0  110  200.0  8.50  2.56   2   3  179.3  65.4  3180    1
8   34.70   89.7   70   81.0  8.20  3.90   2   4  155.7  64.0  1905    0
9   30.40   96.9   75   83.0  9.00  4.30   2   5  165.2  65.0  2320    0
10  16.50  350.0  155  250.0  8.50  3.08   4   3  195.4  74.4  3885    1
11  36.50   85.3   80   83.0  8.50  3.89   2   4  160.6  62.2  2009    0
12  21.50  171.0  109  146.0  8.20  3.22   2   4  170.4  66.9  2655    0
13  19.70  258.0  110  195.0  8.00  3.08   1   3  171.5  77.0  3375    1
14  20.30  140.0   83  109.0  8.40  3.40   2   4  168.8  69.4  2700    0
15  17.80  302.0  129  220.0  8.00  3.00   2   3  199.9  74.0  3890    1
16  14.39  500.0  190  360.0  8.50  2.73   4   3  224.1  79.8  5290    1
17  14.89  440.0  215  330.0  8.20  2.71   4   3  231.0  79.7  5185    1
18  17.80  350.0  155  250.0  8.50  3.08   4   3  196.7  72.2  3910    1
19  16.41  318.0  145  255.0  8.50  2.45   2   3  197.6  71.0  3660    1
20  23.54  231.0  110  175.0  8.00  2.56   2   3  179.3  65.4  3050    1
21  21.47  360.0  180  290.0  8.40  2.45   2   3  214.2  76.3  4250    1
22  16.59  400.0  185    NaN  7.60  3.08   4   3  196.0  73.0  3850    1
23  31.90   96.9   75   83.0  9.00  4.30   2   5  165.2  61.8  2275    0
24  29.40  140.0   86    NaN  8.00  2.92   2   4  176.4  65.4  2150    0
25  13.27  460.0  223  366.0  8.00  3.00   4   3  228.0  79.8  5430    1
26  23.90  133.6   96  120.0  8.40  3.91   2   5  171.5  63.4  2535    0
27  19.73  318.0  140  255.0  8.50  2.71   2   3  215.3  76.3  4370    1
28  13.90  351.0  148  243.0  8.00  3.25   2   3  215.5  78.5  4540    1
29  13.27  351.0  148  243.0  8.00  3.26   2   3  216.1  78.5  4715    1
30  13.77  360.0  195  295.0  8.25  3.15   4   3  209.3  77.4  4215    1
31  16.50  360.0  165  255.0  8.50  2.73   4   3  185.2  69.0  3660    1

In [44]:
X = df['x1']
Y = df['y']
X = sm.add_constant(X)
model = sm.OLS(Y, X).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.772
Model:                            OLS   Adj. R-squared:                  0.765
Method:                 Least Squares   F-statistic:                     101.7
Date:                Fri, 14 Mar 2025   Prob (F-statistic):           3.74e-11
Time:                        14:10:20   Log-Likelihood:                -80.215
No. Observations:                  32   AIC:                             164.4
Df Residuals:                      30   BIC:                             167.4
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         33.7227      1.444     23.355      0.0

b. Construct the analysis-of-variance table and test for significance of regression.  


In [45]:
# Fit the model using the formula interface
b1 = -0.0474
b0 = 33.7227

model_formula = smf.ols('y ~ x1', data=df).fit()
anova_table = sm.stats.anova_lm(model_formula, typ=1)
print(anova_table)

            df      sum_sq     mean_sq           F        PR(>F)
x1         1.0  955.719710  955.719710  101.735668  3.743041e-11
Residual  30.0  281.824378    9.394146         NaN           NaN


In [46]:
#Degree of freedom
df_x1 = 1
df_err = len(df) - 2
print(df_x1, df_err)

1 30


In [47]:
#Sum of Squares
SSR = sum(((b0 + b1*df['x1']) - np.mean(df['y']))**2)
SSE = sum((df['y'] - (b0 + b1*df['x1']))**2)
print(SSR, SSE)

957.3560785796001 281.82930492959997


In [48]:
#Mean Squares
MSR = SSR / df_x1
MSE = SSE / df_err
print(MSR, MSE)

957.3560785796001 9.394310164319998


In [49]:
#Expected Mean Squares
EMSR = MSE + b1**2 * sum((df['x1'] - np.mean(df['x1']))**2)
EMSE = MSE
print(EMSR, EMSE)

966.7461576638698 9.394310164319998


In [50]:
#Test for significant of regression
n = len(df)
F_star = (SSR/1)/(SSE/(n-2))
print("F* = " + str(F_star))
F = stats.f.ppf(q = 1-0.05, dfn=1, dfd = n-2)
p = 1- stats.f.cdf(x = F_star, dfn=1, dfd = n-2)
print(F,F_star, p)

F* = 101.90807646693213
4.170876785766689 101.90807646693213 3.669675674444761e-11


c. What percent of the total variability in gasoline mileage is accounted for by the linear relationship with engine displacement?  
77,2% of the total variability in gasoline mileage is accounted for by the linear relationship with engine displacement

d. Find a 95% CI on the mean gasoline mileage if the engine displacement is 275 in.3  


In [51]:
y_275 = b0 + b1*275
t_val = stats.t.ppf(q = 0.975, df=df_err)
s_y = np.sqrt(MSE*(1/df_err + (275 - np.mean(df['x1']))**2/sum((df['x1'] - np.mean(df['x1']))**2)))

L_ymean = y_275 - t_val * s_y
U_ymean = y_275 + t_val * s_y
print(L_ymean, U_ymean)

19.54080861354697 21.834591386453035


e. Suppose that we wish to predict the gasoline mileage obtained from a car with a 275-in.3 engine. Give a point estimate of mileage. Find a 95% prediction interval on the mileage.  


In [52]:
print("Estimated mileage of 275 displacement is: "+ str(y_275))
F = stats.f.ppf(q = 1-0.05, dfn=2, dfd = n-2)
W = np.sqrt(2*F)
L = y_275 - W * s_y
U = y_275 + W * s_y
print(L, U)

Estimated mileage of 275 displacement is: 20.687700000000003
19.241528813807157 22.13387118619285


f. Compare the two intervals obtained in parts d and e. Explain the difference between them. Which one is wider, and why?  
- CI of yh_mean: Estimates the population mean 𝜇  
- CI of yh (prediction interval): Predicts a new observation y  
- CI of y is wider as it includes both the uncertainty in estimating the mean and variability of data points around the mean

**2.5**
Consider the gasoline mileage data in Table B.3. Repeat Problem 2.4 (parts a, b, and c) using vehicle weight x10 as the regressor variable. Based on a
comparison of the two models, can you conclude that x1 is a better choice of regressor than x10?  
a. Fit a simple linear regression model relating gasoline mileage y (miles per gallon) to vehicle weight x10.  

In [53]:
X = df['x10']
Y = df['y']
X = sm.add_constant(X)
model = sm.OLS(Y,X).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.745
Model:                            OLS   Adj. R-squared:                  0.736
Method:                 Least Squares   F-statistic:                     87.48
Date:                Fri, 14 Mar 2025   Prob (F-statistic):           2.12e-10
Time:                        14:11:31   Log-Likelihood:                -82.047
No. Observations:                  32   AIC:                             168.1
Df Residuals:                      30   BIC:                             171.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         40.8524      2.279     17.926      0.0

b. Construct the analysis-of-variance table and test for significance of regression.  

In [54]:
b0 = 40.8524
b1 = -0.0058 

model2 = smf.ols('y ~ x10', data = df).fit()
anova_table2 = sm.stats.anova_lm(model2, typ=1)
print(anova_table2)

            df      sum_sq     mean_sq          F        PR(>F)
x10        1.0  921.527904  921.527904  87.482346  2.120672e-10
Residual  30.0  316.016184   10.533873        NaN           NaN


In [55]:
#Degree of freedom
df_x1 = 1
df_err = len(df) - 2
print(df_x1, df_err)

1 30


In [56]:
#Sum of Squares
SSR = sum(((b0 + b1*df['x10']) - np.mean(df['y']))**2)
SSE = sum((df['y'] - (b0 + b1*df['x10']))**2)
print(SSR, SSE)

938.0555619999999 317.0447639999999


In [57]:
#Mean Squares
MSR = SSR / df_x1
MSE = SSE / df_err
print(MSR, MSE)

938.0555619999999 10.568158799999996


In [60]:
#Expected Mean Squares
EMSR = MSE + b1**2 * sum((df['x10'] - np.mean(df['x10']))**2)
EMSE = MSE
print(EMSR, EMSE)

947.6603099949999 10.568158799999996


In [61]:
#Expected Mean Squares
EMSR = MSE + b1**2 * sum((df['x1'] - np.mean(df['x1']))**2)
EMSE = MSE
print(EMSR, EMSE)

24.902273769949993 10.568158799999996


In [62]:
#Test for significant of regression
n = len(df)
F_star = (SSR/1)/(SSE/(n-2))
print("F* = " + str(F_star))
F = stats.f.ppf(q = 1-0.05, dfn=1, dfd = n-2)
p = 1- stats.f.cdf(x = F_star, dfn=1, dfd = n-2)
print(F,F_star, p)

F* = 88.7624400572028
4.170876785766689 88.7624400572028 1.7994461476433798e-10


c. What percent of the total variability in gasoline mileage is accounted for by the linear relationship with engine displacement?  
74,5% of the total variability in gasoline mileage is accounted for by the linear relationship with engine displacement.  
**CONCLUSION:** We have 2 models, both fit a simple linear regression model relating gasoline mileage y with model1 to engine displacement x1 while model2 to vehicle weight x10. From the model summary, we can observe that model1 has better R-square value than model2 (0.772 > 0.745), indicating that model1 is better fitted.